include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -105,25 +78,135 @@ template <typename _ExecutionPolicy>
 using __allow_parallel = typename __internal::__policy_traits<::std::decay_t<_ExecutionPolicy>>::__allow_parallel;
 
 template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_vectorization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_vector(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+constexpr auto
+__is_vectorization_preferred()
 {
-    return __internal::__lazy_and(__exec.__allow_vector(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
+    return ::std::conjunction<decltype(::std::decay_t<_ExecutionPolicy>::__allow_vector()),
+                              __internal::__is_random_access_iterator_t<_IteratorTypes...>>();
 }
 
 template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_parallelization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_parallel(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+constexpr auto
+__is_parallelization_preferred()
+{
+    return ::std::conjunction<decltype(::std::decay_t<_ExecutionPolicy>::__allow_parallel()),
+                              __internal::__is_random_access_iterator_t<_IteratorTypes...>>();
+}
+
+//------------------------------------------------------------------------
+// backend selector with tags
+//------------------------------------------------------------------------
+
+#if _ONEDPL_PAR_BACKEND_TBB
+using __par_backend_tag = __tbb_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_OPENMP
+using __par_backend_tag = __omp_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_SERIAL
+using __par_backend_tag = __serial_backend_tag;
+#else
+#    error "Parallel backend was not specified"
+#endif
+
+template <class _IsVector>
+struct __serial_tag
+{
+    using __is_vector = _IsVector;
+};
+
+template <class _IsVector>
+struct __parallel_tag
+{
+    using __is_vector = _IsVector;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+struct __parallel_forward_tag
+{
+    using __is_vector = ::std::false_type;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+template <class _IsVector, class... _IteratorTypes>
+using __tag_type = ::std::conditional_t<
+    __internal::__is_random_access_iterator_v<_IteratorTypes...>, __parallel_tag<_IsVector>,
+    ::std::conditional_t<__is_forward_iterator_v<_IteratorTypes...>, __parallel_forward_tag, __serial_tag<_IsVector>>>;
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::sequenced_policy>,
+    __serial_tag<std::false_type>>
```

Comment:
What's a reason to make the implementation a little bit difference from original one, in Ruslan's prototype (and LLVM upstream) versions?
https://github.com/oneapi-src/oneDPL/blob/rarutyun/tag_dispatching_prototype/include/oneapi/dpl/pstl/execution_impl.h#L123C1-L123C80

Reply:
1) Better to have `__select_backend` function as `constexpr`
2) In the Ruslan's prototype this function receive iterators by value by creating temporary objects. But not all types of our iterators are able to create without params, for example - `permutation_iterator`.

Reply:
@rarutyun, do you insist on "LLVM upstream" approach or you don't mind with Sergey, approach?

Reply:
@rarutyun may I resolve this conversation?

Reply:
I think we agreed on the implementation strategy